### PR TITLE
[release] Add artifact support to release test.

### PR DESCRIPTION
**Pull request by xwjiang2010**
Add artifact support to release test.

Signed-off-by: xwjiang2010 <xwjiang2010@gmail.com>

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/ray_release/schema.json [hunk 1]
```diff
@@ -154,7 +154,10 @@
 				},
 				"long_running": {
 					"type": "boolean"
-				}
+				},
+                "artifact_path": {
+                    "type": "string"
+                }
```
Comment (krfricke) [inline release/ray_release/schema.json:160]: lint please

File: release/long_running_distributed_tests/workloads/pytorch_pbt_failure.py [hunk 2]
```diff
@@ -49,14 +81,86 @@
...
+Expecting to see all trials to reach a minimal checkpoint id of 10 by 1 hour.
+"""
+
+
+class InspectCheckpointIdCallback(Callback):
+    """A callback for asserting the right checkpoint restore behavior.
+
+    Also outputs an artifact that dumps `self._checkpoints` to `/tmp/ckpt_ids.txt`.
+    """
+
```
Comment (krfricke) [inline release/long_running_distributed_tests/workloads/pytorch_pbt_failure.py:105]: Should we still split this into two PRs?

Comment (xwjiang2010) [reply release/long_running_distributed_tests/workloads/pytorch_pbt_failure.py:105]: reverting changes in this file to separate the PRs.

File: release/ray_release/schema.json (hunk 1 above)
Comment (xwjiang2010) [reply release/ray_release/schema.json:141]: I don't know why github ui shows it like this, but in my editor, it looks like
Screen Shot 2023-03-13 at 3 32 26 PM

File: release/ray_release/reporter/artifacts.py [hunk 3]
```diff
@@ -17,6 +17,8 @@
 
 
 class ArtifactsReporter(Reporter):
+    """This is called upon on buildkite runners."""
```
Comment (Yard1) [inline release/ray_release/reporter/artifacts.py:20]:
```suggestion
    """This is called on on buildkite runners."""
```

File: release/ray_release/command_runner/_anyscale_job_wrapper.py [hunk 4]
```diff
@@ -365,6 +378,18 @@ def main(
...
+    parser.add_argument(
+        "--upload-s3-uri",
+        type=str,
+        help="root s3 bucket address to upload stuff",
+        required=False,
+    )
+    parser.add_argument(
+        "--artifact-path",
+        type=str,
+        help="user provided artifact path (on head node)",
```
Comment (Yard1) [inline release/ray_release/command_runner/_anyscale_job_wrapper.py:390]:
```suggestion
        help="user provided artifact path (on head node, must be a single file)",
```

File: release/ray_release/command_runner/command_runner.py [hunk 5]
```diff
@@ -3,16 +3,30 @@
...
+
+    # the directory for runners to dump files to (on buildkite runner instances).
+    # Write to this directory. run_release_tests.sh will ensure that the content
+    # shows up under buildkite job's "Artifacts" UI tab.
+    _DEFAULT_ARTIFACTS_DIR = DEFAULT_ARTIFACTS_DIR
+
+    # the artifact file name put under s3 bucket root.
+    # AnyscalejobWrapper will upload user generated artifact to this path
+    # and AnyscaleJobRunner will then download from there.
+    _USER_GENERATED_ARTIFACT = "user_generated_artifact"
```
Comment (Yard1) [inline release/ray_release/command_runner/command_runner.py:22]: can we also move `result_output_json` and `metrics_output_json` (with consistent capitalization)?

**Review by Yard1 — COMMENTED** [review]
Thanks!

**Comment by Yard1** [comment]
can you also run the release test if you haven't done that already?

**Comment by xwjiang2010** [comment]
running here: https://buildkite.com/ray-project/release-tests-pr/builds/30828

File: release/ray_release/command_runner/command_runner.py (hunk 5 above)
Comment (xwjiang2010) [reply release/ray_release/command_runner/command_runner.py:22]: done. @Yard1 just wanted result_output_json and metrics_output_json  to be moved to class var.

File: release/ray_release/command_runner/command_runner.py [hunk 6]
```diff
@@ -21,6 +21,11 @@ class CommandRunner(abc.ABC):
     # and AnyscaleJobRunner will then download from there.
     _USER_GENERATED_ARTIFACT = "user_generated_artifact"
 
+    # the directory path where output json will
```
Comment (Yard1) [inline release/ray_release/command_runner/command_runner.py:24]: cuts off

**Comment by xwjiang2010** [comment]
Screen Shot 2023-03-14 at 1 47 55 PM

**Review by Yard1 — APPROVED** [review]

**Review by krfricke — APPROVED** [review]
LGTM

**Pull request merged**